numpy/_core/src/umath/ufunc_type_resolution.c

```diff
@@ -429,6 +429,31 @@ PyUFunc_NegativeTypeResolver(PyUFuncObject *ufunc,
     return ret;
 }
 
+/*
+ * This function applies special type resolution rules for the 'sign' ufunc.
+ * 'sign' converts timedelta64 to float64, so isn't covered by the simple
+ * unary type resolution.
+ *
+ * Returns 0 on success, -1 on error.
+ */
+NPY_NO_EXPORT int
+PyUFunc_SignTypeResolver(PyUFuncObject *ufunc,
+                         NPY_CASTING casting,
+                         PyArrayObject **operands,
+                         PyObject *type_tup,
+                         PyArray_Descr **out_dtypes)
+{
+    if (PyArray_DESCR(operands[0])->type_num == NPY_TIMEDELTA) {
+        out_dtypes[0] = NPY_DT_CALL_ensure_canonical(PyArray_DESCR(operands[0]));
```

Comment:
I guess it's practically impossible, but I think this call can actually fail.
Maybe I should actually fix this, so that it can't fail for the simple dtypes (i.e. the simple ones can actually just return the `singleton` here, but they create a new dtype right now).

Other than this nit (and I think it's probably wrong in at least one other place, so I am happy to change `_ensure_canonical` here for the simple dtypes), it looks good to me.

@mhvk I am fine with putting this in, don't see much chance of downstream annoyance. I'll note that astropy asked before to not merge weirder things while the rc is in progress because it can be confusing in CI (i.e. you expect the change to be part of the rc release, but it's not).
But I am fine with just merging either way.